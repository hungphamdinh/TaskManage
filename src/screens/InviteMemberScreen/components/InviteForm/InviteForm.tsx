import React, { useState, useEffect } from "react";
import { View, Image } from "react-native";
import styles from "./styles";
import { User } from "../../../../services/model/User";
import { AppText, AppButton } from "../../../../components";
import { Colors, Fonts } from "../../../../themes";
import {
  TouchableOpacity,
  TextInput,
  FlatList,
} from "react-native-gesture-handler";
import { Ionicons } from "@expo/vector-icons";
import { KeyboardAwareScrollView } from "react-native-keyboard-aware-scroll-view";
import { useSelector } from "react-redux";
import ReduxState from "../../../../redux/ReduxState";
import { strings } from "../../../../languages";
import {
  getUsers,
  searchUsers,
  addUser,
  clearLocalUser,
} from "../../../../redux/user/reducer/usersById";
import {
  sendInvitation, clearInvitationSend,
} from "../../../../redux/invitation/action/invitationSend";
import { SendInvitationRequest } from "../../../../services/model/request/Invitation";

const AddForm = ({
  dispatch,
  user,
  onNavigate,
}: {
  dispatch: any;
  user: User;
  onNavigate: Function;
}) => {
  const { users, usersLocal } = useSelector(
    (state: ReduxState) => state.usersById
  );
  const { taskDetail } = useSelector((state: ReduxState) => state.taskDetail);
  const { response } = useSelector((state: ReduxState) => state.invitationSend);
  const [name, setName] = useState("");
  useEffect(() => {
    if (users.length == 0) {
      dispatch(
        getUsers({
          id: user.id,
        })
      );
    }
    return () => {
      dispatch(clearInvitationSend());
      dispatch(clearLocalUser());
    }
  }, []);

  useEffect(() => {
    if (response) {
      onNavigate();
    }
  }, [response]);
  const _onChangeTaskName = (value: any) => {
    setName(value);
    dispatch(searchUsers(value));
  };

  const _onPressDone = () => {
    let arr: Array<SendInvitationRequest> = [];
    usersLocal.map((item: User) => {
      if (item.isActive) {
        arr.push({
          userId: user.id,
          taskId: taskDetail.id,
          userName: user.name,
          receiverId: item.id,
        });
      }
    });
    dispatch(sendInvitation(arr));
  };

  const _keyExtractor = (item: any, index: number) => index.toString();

  const _renderItem = ({ item, index }: { item: User; index: number }) => (
    <Item index={index} item={item} onPress={_onPressItem} />
  );

  const _onPressItem = (item: User, index: number) => {
    dispatch(addUser(item));
  };
  return (
    <KeyboardAwareScrollView contentContainerStyle={styles.container}>
      <View style={styles.inputTask}>
        <TextInput
          value={name}
          onChangeText={_onChangeTaskName}
          placeholder={strings.add_member_screen.search}
        />
        <View style={styles.icSearch}>
          <Ionicons
            name="search-outline"
            size={20}
            color={Colors.appGrayColor}
          />
        </View>
      </View>
      <View style={styles.teamContainer}>
        <FlatList
          data={name !== "" ? usersLocal : users}
          renderItem={_renderItem}
          keyExtractor={_keyExtractor}
        />
      </View>

      <View style={styles.buttonDone}>
        <AppButton text={"Done"} onPress={_onPressDone} />
      </View>
    </KeyboardAwareScrollView>
  );
};

export default AddForm;

const Item = ({
  item,
  index,
  onPress,
}: {
  item: User;
  index: number;
  onPress: Function;
}) => {
  const _onPress = () => {
    onPress(item, index);
  };
  return (
    <TouchableOpacity style={styles.itemContainer} onPress={_onPress}>
      <View style={styles.profileContainer}>
        <View style={styles.imageProfileContainer}>
          <Image style={styles.imageMember} source={{ uri: item.profile }} />
        </View>
        <View style={styles.roleContainer}>
          <AppText text={item.name} bold size={Fonts.size.large} />
          <AppText
            style={styles.marginTopSmall}
            text={item.role}
            color={Colors.appGrayColor}
          />
        </View>
      </View>
      <Ionicons
        name="checkmark-outline"
        size={20}
        color={item.isActive ? Colors.appPrimaryColor : Colors.appWhite}
      />
    </TouchableOpacity>
  );
};
