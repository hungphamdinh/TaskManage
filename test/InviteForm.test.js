import React from "react";
import { cleanup } from "@testing-library/react-native";
import * as reactRedux from "react-redux";
import InviteForm from "../src/screens/team/InviteMemberScreen/components/InviteForm/InviteForm";
import Enzyme, { shallow } from "enzyme";
import Adapter from "enzyme-adapter-react-16";
import toJson from "enzyme-to-json";

jest.mock("react-native-keyboard-aware-scroll-view", () => {
  const KeyboardAwareScrollView = ({ children }) => children;
  return { KeyboardAwareScrollView };
});
Enzyme.configure({ adapter: new Adapter() });
const useSelectorMock = jest.spyOn(reactRedux, "useSelector");
const useDispatchMock = jest.spyOn(reactRedux, "useDispatch");

const setState = jest.fn();
const useStateSpy = jest.spyOn(React, "useState");
useStateSpy.mockImplementation((init) => [init, setState]);

let wrapper;
const userMock = {
  id: "0",
  googleUserId: "0",
  name: "Phạm Đình HƯng",
  mail: "abc@gmail.com",
  role: "Deb",
  profile: "",
  members: [],
};

const mockUsers = [
  {
    id: "0",
    googleUserId: "0",
    name: "Phạm Đình HƯng",
    mail: "abc@gmail.com",
    role: "Deb",
    profile: "",
    members: [],
  },
  {
    id: "1",
    googleUserId: "0",
    name: "Hello",
    mail: "abc@gmail.com",
    role: "Deb",
    profile: "",
    members: [],
  },
  {
    id: "2",
    googleUserId: "0",
    name: "Hihi",
    mail: "abc@gmail.com",
    role: "Deb",
    profile: "",
    members: [],
  },
];
const userAction = {
  params: {
    id: "abc",
    taskId: "abcd12312312",
  },
  type: "GET_USERS",
};

const userActionSearchMember = {
  params: {
    id: "abc",
    taskId: "abcd12312312",
  },
  name: "somenewpassword",
  type: "GET_USERS_ACTION_SEARCH_MEMBER"
};
const dummyDispatch = jest.fn();
useDispatchMock.mockReturnValue(dummyDispatch);
useSelectorMock.mockReturnValue({ user: userMock, users: mockUsers });

describe("Invite Form", () => {
  beforeEach(() => {
    useSelectorMock.mockClear();
    useDispatchMock.mockClear();
    wrapper = shallow(<InviteForm user={userMock} dispatch={dummyDispatch} />);
  });

  test("InviteForm renders correctly", () => {
    const tree = wrapper;
    expect(toJson(tree)).toMatchSnapshot();
  });

  it("Search Input props", () => {
    const props = {
      testID: "searchInput",
      value: "",
      onChangeText: expect.any(Function),
      placeholder: "Search...",
      allowFontScaling: true,
      rejectResponderTermination: true,
      underlineColorAndroid: "transparent",
    };
    expect(wrapper.find({ testID: "searchInput" }).props()).toEqual(props);

    // expect(wrapper.props().product).toEqual(product);
  });

  it("Search Input value", () => {
    wrapper
      .find({ testID: "searchInput" })
      .simulate("changeText", "somenewpassword");
    wrapper.update();
    expect(wrapper.find({ testID: "searchInput" }).prop("value")).toEqual(
      "somenewpassword"
    );
    expect(dummyDispatch).toHaveBeenCalledTimes(1);
    expect(dummyDispatch).toHaveBeenCalledWith(userActionSearchMember);

  });

  it("should flatlist render Correct", () => {
    const flatListWrapper = wrapper.find("FlatList").props();
    const key = flatListWrapper.keyExtractor(3, 10);
    const listData = flatListWrapper.data;
    expect(key).toEqual("10");
    expect(listData.length).toEqual(mockUsers.length);
  });

  // it("should mock Dispatch", () => {
  //   // expect(dummyDispatch).toHaveBeenCalledWith(userAction);
  //   expect(dummyDispatch).toHaveBeenCalledTimes(2);
  //   expect(dummyDispatch).toHaveBeenCalledWith(userAction);
  // });
});

afterEach(cleanup);

// it("InviteForm Function", () => {
//   const instanceOf = renderer.create(<InviteForm />);
//   console.log(instanceOf);
//   // instanceOf._onFocus();
//   expect(instanceOf.state.isFocus).toEqual(true);
// });
