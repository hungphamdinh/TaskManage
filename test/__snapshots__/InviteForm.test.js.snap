// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Invite Form InviteForm renders correctly 1`] = `
<KeyboardAwareScrollView
  contentContainerStyle={
    Object {
      "flexGrow": 1,
      "marginHorizontal": 25,
    }
  }
>
  <View
    style={
      Object {
        "backgroundColor": "#fff",
        "borderRadius": 10,
        "flexDirection": "row",
        "justifyContent": "space-between",
        "marginTop": 20,
        "padding": 15,
      }
    }
  >
    <Component
      allowFontScaling={true}
      onChangeText={[Function]}
      placeholder="Search..."
      rejectResponderTermination={true}
      testID="searchInput"
      underlineColorAndroid="transparent"
      value=""
    />
    <View
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#F0F4FD",
          "borderRadius": 7.5,
          "height": 30,
          "justifyContent": "center",
          "width": 30,
        }
      }
    >
      <Icon
        allowFontScaling={false}
        color="#919BB3"
        name="search-outline"
        size={20}
      />
    </View>
  </View>
  <View
    style={Object {}}
  >
    <FlatList
      data={
        Array [
          Object {
            "googleUserId": "0",
            "id": "0",
            "mail": "abc@gmail.com",
            "members": Array [],
            "name": "Phạm Đình HƯng",
            "profile": "",
            "role": "Deb",
          },
          Object {
            "googleUserId": "0",
            "id": "1",
            "mail": "abc@gmail.com",
            "members": Array [],
            "name": "Hello",
            "profile": "",
            "role": "Deb",
          },
          Object {
            "googleUserId": "0",
            "id": "2",
            "mail": "abc@gmail.com",
            "members": Array [],
            "name": "Hihi",
            "profile": "",
            "role": "Deb",
          },
        ]
      }
      disableVirtualization={false}
      horizontal={false}
      initialNumToRender={10}
      keyExtractor={[Function]}
      maxToRenderPerBatch={10}
      numColumns={1}
      onEndReachedThreshold={2}
      removeClippedSubviews={false}
      renderItem={[Function]}
      scrollEventThrottle={50}
      updateCellsBatchingPeriod={50}
      windowSize={21}
    />
  </View>
  <View
    style={
      Object {
        "flexGrow": 1,
        "justifyContent": "flex-end",
      }
    }
  >
    <AppButton
      onPress={[Function]}
      text="Done"
    />
  </View>
</KeyboardAwareScrollView>
`;
